File: mobx/src/main/java/com/ivansadovyi/mobx/When.java
package com.ivansadovyi.mobx;

public class When implements Disposable {

	private Disposable disposable;

	public When(final WhenPredicate predicate, final Runnable sideEffect) {
		disposable = new Reaction<>(
				new ReactionDataCallback<Boolean>() {
					@Override
					public Boolean getData() {
						return predicate.test();
					}
				},
				new ReactionSideEffectCallback<Boolean>() {
					@Override
					public void onReaction(Boolean data) {
						if (data) {
							dispose();
							sideEffect.run();
						}
					}
				}
		);
	}

	@Override
	public void dispose() {
		disposable.dispose();
	}
}

File: mobx/src/main/java/com/ivansadovyi/mobx/When.kt
package com.ivansadovyi.mobx

typealias WhenPredicate = () -> Boolean
typealias WhenSideEffectCallback = () -> Unit

class When(predicate: WhenPredicate, sideEffectCallback: WhenSideEffectCallback) : Disposable {

	private var reactionDisposable: Disposable? = null

	init {
		this.reactionDisposable = reaction(
			dataCallback = predicate,
			sideEffectCallback = {
				if (it) {
					sideEffectCallback()
					this.reactionDisposable?.dispose()
				}
			},
			fireImmediately = true
		)
	}

	override fun dispose() {
		this.reactionDisposable?.dispose()
	}
}
